src/resolve/resolved-dns-transaction.c

```diff
@@ -509,8 +508,68 @@ static int on_stream_complete(DnsStream *s, int error) {
         return 0;
 }
 
-static int dns_transaction_open_tcp(DnsTransaction *t) {
+static int on_stream_complete(DnsStream *s, int error) {
+        DnsTransaction *t, *n;
+        int r = 0;
+
+        if (ERRNO_IS_DISCONNECT(error) && s->protocol != DNS_PROTOCOL_LLMNR) {
+                usec_t usec;
+
+                log_debug_errno(error, "Connection failure for DNS TCP stream: %m");
+
+                if (s->transactions) {
+                        t = s->transactions;
+                        assert_se(sd_event_now(t->scope->manager->event, clock_boottime_or_monotonic(), &usec) >= 0);
+                        dns_server_packet_lost(t->server, IPPROTO_UDP, t->current_feature_level, usec - t->start_usec);
+                }
+        }
+
+        LIST_FOREACH_SAFE(transactions_by_stream, t, n, s->transactions)
+                if (error != 0)
+                        on_transaction_stream_error(t, error);
+                else if (DNS_PACKET_ID(s->read_packet) == t->id)
+                        r = on_dns_stream_packet(t, s->read_packet);
+
+        dns_stream_unref(s);
+
+        return r;
+}
+
+static int on_stream_packet(DnsStream *s) {
+        _cleanup_(dns_packet_unrefp) DnsPacket *p = NULL;
+        int r = 0;
+        DnsTransaction *t, *ct = NULL;
+
+        /* Take ownership of packet to be able to receive new packets */
+        p = s->read_packet;
+        s->read_packet = NULL;
+        s->n_read = 0;
+
+        /* Find corresponding transaction for this answer */
+        LIST_FOREACH(transactions_by_stream, t, s->transactions)
+                if (DNS_PACKET_ID(p) == t->id) {
+                        ct = t;
+                        break;
+                }
```

Comment:
you can use `hashmap_get(s->manager->dns_transactions, UINT_TO_PTR(DNS_PACKET_ID(p))` to get the transaction for the specified ID directly in O(1), no need to search linearly here